### PR TITLE
Issue 88: 5. Add network administrator

### PR DESCRIPTION
**Pull request by LourensVeen**
This is the skeleton for the network administrator, which is going to be in charge of managing container networking. <Usual stacking/merge disclaimer>.

### CODE REVIEW

File: mahiru/components/domain_administrator.py [hunk 1]
```diff
@@ -61,12 +62,16 @@ class PlainDockerDA(IDomainAdministrator):
...
+    def __init__(
+            self, network_administrator: INetworkAdministrator,
+            site_rest_client: SiteRestClient) -> None:
         """Create a PlainDockerDA.
 
         Args:
+            network_administrator: Network administrator to use.
             site_rest_client: Client to use for downloading images.
         """
+        self._network_administrator = network_administrator
```
Comment (svenvanderburg) [inline mahiru/components/domain_administrator.py:74]: I guess our good old friend `PlainDockerDA` will start using his new powers in the next chapter of this book you are writing in Github?

File: mahiru/components/network_administrator.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        """Connect a local network namespace to a set of inputs.
+
+        Args:
+            job_id: Job id for future reference.
+            inputs: Assets to connect to, indexed by input name.
+            network_namespace: Namespace to create network interfaces
+                    inside of.
+
+        Return:
+            (nets, remaining) where nets contains a host IP for each
```
Comment (svenvanderburg) [inline mahiru/components/network_administrator.py]: I guess `nets` is also indexed by `input name`? Maybe add this to docstring.

File: mahiru/components/network_administrator.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            assets from inputs that we could not connect to.
+        """
+        raise NotImplementedError()
+
+    def disconnect_inputs(self, job_id: int, inputs: Dict[str, Asset]) -> None:
+        """Disconnect inputs and free resources.
+
+        Args:
+            job_id: Job id of job to disconnect.
+            inputs: Input assets for this job.
```
Comment (svenvanderburg) [inline mahiru/components/network_administrator.py]:
```suggestion
            inputs: Input assets for this job. Indexed by input name
```

File: mahiru/definitions/interfaces.py [hunk 4]
```diff
@@ -213,3 +217,69 @@ def execute_step(
...
+
+class INetworkAdministrator:
+    """Manages network resources for a site.
+
+    The "domain" in the name is a network administration domain,
+    containing (virtual) networks and potentially (virtual)
+    programmable networking hardware, etc. in which workflows are
+    executed. Classes implementing this interface manage
+    ("administrate") these resources in the domain to help implement
+    workflow execution.
```
Comment (svenvanderburg) [inline mahiru/definitions/interfaces.py]: This explanation seems to refer to something else?

File: mahiru/components/network_administrator.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        Args:
+            conn_id: Connection id for this connection.
+            network_namespace: PID of the network namespace to create
+                    the endpoint inside of.
+            request: Connection request from the client side.
+
+        Return:
+            A connection info object to send back to the client.
+        """
+        raise NotImplementedError()
```
Comment (svenvanderburg) [inline mahiru/components/network_administrator.py:49]: Also these methods will be implemented in a later PR I assume?

**Review by svenvanderburg — APPROVED** [review]
Looks good. I am looking forward to the next chapter... I left a few minor comments.

File: mahiru/components/domain_administrator.py (hunk 1 above)
Comment (LourensVeen) [reply mahiru/components/domain_administrator.py:74]: Yep! Stay tuned...

File: mahiru/components/network_administrator.py (hunk 2 above)
Comment (LourensVeen) [reply mahiru/components/network_administrator.py]: Ah yes, that's what "for each [...] input" was intended to convey, but you're right that it should be more explicit. Will fix.

File: mahiru/definitions/interfaces.py (hunk 4 above)
Comment (LourensVeen) [reply mahiru/definitions/interfaces.py]: It explains the name `DomainAdministrator` by describing what we mean by "domain" and "administrate"? I'm not sure what I could improve here.

File: mahiru/components/network_administrator.py (hunk 5 above)
Comment (LourensVeen) [reply mahiru/components/network_administrator.py:49]: Yep!

File: mahiru/definitions/interfaces.py (hunk 4 above)
Comment (svenvanderburg) [reply mahiru/definitions/interfaces.py]: This is the description for the `NetworkAdministrator` not the `DomainAdministrator`😉

Comment (LourensVeen) [reply mahiru/definitions/interfaces.py]: Doh!

**Pull request merged**